File: lua/scopes.lua
-- Labeled scope test cases for Lua. See SCOPE_TEST_SPEC.md at repo root.
-- N/A for Lua: S09 (require returns values, no `import as`), S12 (metatable
-- emulation rather than real statics), S14 (tables, not namespaces).

local M = {}

local module2 = require("module2")              -- same-pattern cross-module access
local scopes_ns = require("scopes_ns")          -- S14 stand-in (table member)

-- S04 def / S05 write target. Using a module-local means it's "private" to
-- this file; _G.GlobalVar is the true "global" for cross-module scenarios.
M.MODULE_VAR = "mod-initial"                    -- S04.def

function M.s01_local()
    local local_a = "S01.local"                 -- S01.def
    print(local_a)                              -- S01.read
end

function M.s02_closure_read()
    local outer_a = "S02.outer"                 -- S02.outer.def
    local inner = function() print(outer_a) end -- S02.inner.read (upvalue)
    inner()
end

function M.s03_closure_write()
    local counter = 0                           -- S03.outer.def
    local bump = function() counter = counter + 1 end  -- S03.inner.write (upvalue)
    bump(); bump()
    return counter                              -- S03.outer.read
end

function M.s05_same_module_write()
    M.MODULE_VAR = "rotated"                    -- S05.write
    print(M.MODULE_VAR)                         -- S05.read
end

function M.s06_cross_read()
    return module2.MODULE2_GLOBAL               -- S06.read
end

function M.s07_cross_write()
    module2.set_global("S07")                   -- S07.write (indirect)
end

function M.s08_shadowing()
    local MODULE_VAR = "shadowed"               -- S08.shadow.def — local shadows M.MODULE_VAR
    print(MODULE_VAR)                           -- S08.shadow.read
end

-- S10 re-export stand-in: scopes_reexport defines a value, scopes_ns
-- re-exports it via its own module table. Reader goes through scopes_ns.
function M.s10_reexport()
    print(scopes_ns.re_exported_value)          -- S10.consumer.read
end

-- S11/S13 — metatable-based "class" inheritance.
local ScopeBase = {}
ScopeBase.__index = ScopeBase
function ScopeBase.new(x)
    return setmetatable({ x = x }, ScopeBase)   -- S11.instance.def (field x)
end
function ScopeBase:read_instance(x)
    return x + self.x                           -- S11.param.read + S11.instance.read
end

local ScopeDerived = setmetatable({}, { __index = ScopeBase })
ScopeDerived.__index = ScopeDerived
function ScopeDerived.new()
    return setmetatable({ x = 7 }, ScopeDerived)
end
function ScopeDerived:read_inherited()
    return self:read_instance(100)              -- S13.derived.read — routes to ScopeBase
end

function M.run_scope_demo()
    M.s01_local()
    M.s02_closure_read()
    print(M.s03_closure_write())
    M.s05_same_module_write()
    print(M.s06_cross_read())
    M.s07_cross_write()
    M.s08_shadowing()
    M.s10_reexport()
    print(ScopeBase.new(42):read_instance(100))
    print(ScopeDerived.new():read_inherited())
end

return M

File: ocaml/scopes.ml
(* Labeled scope test cases for OCaml. See SCOPE_TEST_SPEC.md. *)

let module_var = ref "mod-initial"                    (* S04.def — mutable ref *)

let s01_local () =
  let local_a = "S01.local" in                        (* S01.def *)
  print_endline local_a                               (* S01.read *)

let s02_closure_read () =
  let outer_a = "S02.outer" in                        (* S02.outer.def *)
  let inner () = print_endline outer_a in             (* S02.inner.read *)
  inner ()

let s03_closure_write () =
  let counter = ref 0 in                              (* S03.outer.def *)
  let bump () = incr counter in                       (* S03.inner.write *)
  bump (); bump ();
  !counter                                            (* S03.outer.read *)

let s05_same_module_write () =
  module_var := "rotated";                            (* S05.write *)
  print_endline !module_var                           (* S05.read *)

let s08_shadowing () =
  let module_var = "shadowed" in                      (* S08.shadow.def *)
  print_endline module_var                            (* S08.shadow.read *)

(* S11: record field vs function param. *)
type scope_base = { x : int }
let read_instance x b = x + b.x                       (* S11.param.read + S11.instance.read *)

module Ns = struct
  type widget = { label : string }                    (* S14.Widget.def *)
  let mk_widget label = { label }
end

let s14_qualified () = (Ns.mk_widget "hi").label      (* S14.read *)

let run_scope_demo () =
  s01_local ();
  s02_closure_read ();
  Printf.printf "counter=%d\n" (s03_closure_write ());
  s05_same_module_write ();
  s08_shadowing ();
  Printf.printf "s11=%d\n" (read_instance 100 { x = 42 });
  Printf.printf "s14=%s\n" (s14_qualified ())
